pkg/patterns/declarative/pkg/watch/dynamic.go

```diff
@@ -50,32 +54,51 @@ func (dw *dynamicWatch) newDynamicClient(gvk schema.GroupVersionKind) (dynamic.R
 
 // Add registers a watch for changes to 'trigger' filtered by 'options' to raise an event on 'target'
 func (dw *dynamicWatch) Add(trigger schema.GroupVersionKind, options metav1.ListOptions, target metav1.ObjectMeta) error {
-	log := log.Log
-
 	client, err := dw.newDynamicClient(trigger)
 	if err != nil {
 		return fmt.Errorf("creating client for (%s): %v", trigger.String(), err)
 	}
 
+	go func() {
+		lastResourceVersion := ""
+		for {
+			lastResourceVersion = dw.watchUntilClosed(lastResourceVersion, client, trigger, options, target)
+
+			time.Sleep(WatchDelay)
+		}
+	}()
+
+	return nil
+}
+
+func (dw *dynamicWatch) watchUntilClosed(lastResourceVersion string, client dynamic.ResourceInterface, trigger schema.GroupVersionKind, options metav1.ListOptions, target metav1.ObjectMeta) string {
+	log := log.Log
+
+	options.ResourceVersion = lastResourceVersion
 	events, err := client.Watch(options)
 
 	if err != nil {
-		log.WithValues("kind", trigger.String()).WithValues("namespace", target.Namespace).WithValues("labels", options.LabelSelector).Error(err, "adding watch to dynamic client")
-		return fmt.Errorf("adding watch to dynamic client: %v", err)
+		log.WithValues("kind", trigger.String()).WithValues("namespace", target.Namespace).WithValues("labels", options.LabelSelector).WithValues("lastResourceVersion", lastResourceVersion).Error(err, "adding watch to dynamic client")
+		return lastResourceVersion
 	}
 
-	log.WithValues("kind", trigger.String()).WithValues("namespace", target.Namespace).WithValues("options", options.String()).V(2).Info("watch registered")
-	eventChan := events.ResultChan()
+	accessor := meta.NewAccessor()
 
-	go func() {
-		for clientEvent := range eventChan {
-			if clientEvent.Object == nil || clientEvent.Type == watch.Added {
-				continue
-			}
-			log.WithValues("type", clientEvent.Type).WithValues("kind", trigger.String()).Info("broadcasting event")
-			dw.events <- event.GenericEvent{Object: clientEvent.Object, Meta: &target}
+	log.WithValues("kind", trigger.String()).WithValues("namespace", target.Namespace).WithValues("labels", options.LabelSelector).WithValues("lastResourceVersion", lastResourceVersion).Info("watch began")
+	for clientEvent := range events.ResultChan() {
+		version, err := accessor.ResourceVersion(clientEvent.Object)
+		if err == nil {
+			lastResourceVersion = version
 		}
-	}()
 
-	return nil
+		if clientEvent.Object == nil || clientEvent.Type == watch.Added {
```

Comment:
It's a little surprising that we always drop watch.Added.  Should we send it and let the client filter?

Reply:
agreed! dropped.